File: lib/rayz/group.rb
require_relative "shape"

module Rayz
  class Group < Shape
    attr_reader :children

    def initialize
      super
      @children = []
    end

    def add_child(shape)
      @children << shape
      shape.parent = self
    end

    def empty?
      @children.empty?
    end

    def include?(shape)
      @children.include?(shape)
    end

    def local_intersect(local_ray)
      # Collect intersections from all children
      intersections = []
      @children.each do |child|
        xs = child.intersect(local_ray)
        intersections.concat(xs)
      end
      # Sort by t value
      intersections.sort_by(&:t)
    end

    def local_normal_at(local_point, hit = nil)
      # Groups have no surface of their own, so this should never be called
      raise "Groups have no surface and cannot have normals computed"
    end

    # Override includes? to recursively check children
    def includes?(shape)
      @children.any? { |child| child.includes?(shape) }
    end
  end
end

File: python/rayz/group.py
from __future__ import annotations

from rayz.shape import Shape
from rayz.tuple import Vector


class Group(Shape):
    def __init__(self) -> None:
        super().__init__()
        self.children: list[Shape] = []
        self._bounds_cache = None

    def add_child(self, shape: Shape) -> None:
        self.children.append(shape)
        shape.parent = self
        self.invalidate_bounds_cache()

    def invalidate_bounds_cache(self) -> None:
        self._bounds_cache = None
        super().invalidate_bounds_cache()

    def local_intersect(self, ray) -> list:
        if self.children and not self.bounds().intersects(ray):
            return []
        xs = []
        for child in self.children:
            xs.extend(child.intersect(ray))
        return sorted(xs, key=lambda i: i.t)

    def local_normal_at(self, point, hit=None) -> Vector:
        raise RuntimeError("Groups have no surface normals")

    def includes(self, shape) -> bool:
        return any(child.includes(shape) for child in self.children)

    def bounds(self):
        if self._bounds_cache is None:
            from rayz.bounds import Bounds

            b = Bounds()
            for child in self.children:
                child_bounds = child.bounds().transform(child.transform)
                b = b.merge(child_bounds)
            self._bounds_cache = b
        return self._bounds_cache
